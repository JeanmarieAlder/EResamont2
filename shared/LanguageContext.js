import { createContext } from "react";

export const LanguageContext = createContext();
// 1 is French
// 2 is Italian
// 3 is English
// 4 is Deutsch
