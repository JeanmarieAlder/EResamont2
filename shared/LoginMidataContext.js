import React, { useState, useEffect, useContext } from "react";
import { ToastAndroid } from "react-native";
import * as AppAuth from "expo-app-auth";
import * as SecureStore from "expo-secure-store";
import requestGetMidata from "../utils/requestGetMidata";

export const Auth2Context = React.createContext();
export const useAuth2 = () => useContext(Auth2Context);
export const Auth2Provider = ({ children }) => {
  const StorageKey = "EResamont2MidataOAuthKey";
  const config = {
    issuer: "https://test.midata.coop/fhir/metadata",
    clientId: "eresamont2-test",
    serviceConfiguration: {
      authorizationEndpoint: "https://test.midata.coop/authservice",
      tokenEndpoint: "https://test.midata.coop/v1/token",
      revocationEndpoint: "https://test.midata.coop/v1/token"
    },
    redirectUrl: AppAuth.OAuthRedirect + ":/oauthredirect"
  };

  const [authState, setAuthState] = useState(null);
  const [userName, setUserName] = useState(null);

  useEffect(() => {
    console.log("LoginMidataContext: useEffect()");
    (async () => {
      let cachedAuth = await getCachedAuthAsync();
      if (cachedAuth && !authState) {
        setAuthState(cachedAuth);
      }
    })();
  }, []);

  const signInAsync = async () => {
    console.log("signInAsync()");
    let authState = await AppAuth.authAsync(config);
    await cacheAuthAsync(authState);
    console.log("signInAsync", authState);
    setAuthState(authState);
    return authState;
  };

  const signOutAsync = async ({ accessToken }) => {
    try {
      await AppAuth.revokeAsync(config, {
        token: accessToken,
        isClientIdProvided: true
      });
      await SecureStore.deleteItemAsync(StorageKey);
      console.log("sign out completed");
      ToastAndroid.show("Sign out completed", ToastAndroid.SHORT);
      setAuthState(null);
      return null;
    } catch (e) {
      console.error(`Failed to revoke token: ${e.message}`);
    }
  };

  const cacheAuthAsync = async authState => {
    return await SecureStore.setItemAsync(
      StorageKey,
      JSON.stringify(authState)
    );
  };

  const getCachedAuthAsync = async () => {
    let value = await SecureStore.getItemAsync(StorageKey);
    let authState = JSON.parse(value);
    console.log("getCachedAuthAsync", authState);
    if (authState) {
      if (checkIfTokenExpired(authState)) {
        return refreshAuthAsync(authState);
      } else {
        return authState;
      }
    }
    return null;
  };

  const checkIfTokenExpired = ({ accessTokenExpirationDate }) => {
    return new Date(accessTokenExpirationDate) < new Date();
  };

  const refreshAuthAsync = async ({ refreshToken }) => {
    let authState = await AppAuth.refreshAsync(config, refreshToken);
    console.log("refreshAuth", authState);
    await cacheAuthAsync(authState);
    return authState;
  };

  const getLoggedUserName = async () => {
    if (authState) {
      let response = await requestGetMidata(
        "https://test.midata.coop/fhir/Patient/" +
          authState.additionalParameters.patient,
        authState,
        signInAsync
      );
      if (response) {
        setUserName(response.name[0].given[0]);
        return true;
      }
    }
    return false;
  };

  return (
    <Auth2Context.Provider
      value={{
        authState,
        userName,
        signInAsync,
        signOutAsync,
        cacheAuthAsync,
        getCachedAuthAsync,
        checkIfTokenExpired,
        refreshAuthAsync,
        getLoggedUserName
      }}
    >
      {children}
    </Auth2Context.Provider>
  );
};
