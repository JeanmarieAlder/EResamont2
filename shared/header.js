import React from "react";
import { StyleSheet, Text, View } from "react-native";
import { MaterialIcons } from "@expo/vector-icons";

export default function Header({ title, navigation }) {
  const openMenu = () => {
    navigation.openDrawer();
  };

  return (
    <View style={styles.header}>
      <MaterialIcons
        name="menu"
        size={28}
        onPress={openMenu}
        style={styles.icon}
      />
      <View>
        <Text style={styles.headerText}>{title}</Text>
      </View>
    </View>
  );
}

const styles = StyleSheet.create({
  header: {
    flexDirection: "row",
    justifyContent: "flex-end",
    flex: 1
  },
  headerText: {
    fontWeight: "bold",
    fontSize: 20,
    color: "#fff",
    alignItems: "flex-end",
    letterSpacing: 1,
    marginLeft: 10
  },
  icon: { color: "#fff" }
});
