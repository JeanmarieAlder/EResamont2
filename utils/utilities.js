export default class utilities {
  static findLanguageIndex(data, language) {
    let index = data.findIndex(item => item.language.id === language);
    return index;
  }
}
