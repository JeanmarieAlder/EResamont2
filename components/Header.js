import React, { useContext } from "react";
import { StyleSheet, Text, View } from "react-native";
import { MaterialIcons } from "@expo/vector-icons";
import { globalStyles } from "../styles/global";
import { LanguageContext } from "../shared/LanguageContext";
import utilities from "../utils/utilities";

export default function Header({ title, navigation }) {
  const { language, setLanguage } = useContext(LanguageContext);

  const openMenu = () => {
    navigation.openDrawer();
  };
  let findTitle = title => {
    if (title == "Home" || title == "test title") {
      return title;
    } else {
      return title[utilities.findLanguageIndex(title, language)].title;
    }
  };

  return (
    <View style={localStyles.header}>
      <MaterialIcons
        name="menu"
        size={24}
        onPress={openMenu}
        style={globalStyles.headerIcon}
        testID={"header-button-menu"}
      />
      <View>
        <Text style={globalStyles.headerText}>{title && findTitle(title)}</Text>
      </View>
    </View>
  );
}

const localStyles = StyleSheet.create({
  header: {
    flexDirection: "row"
    // justifyContent: "flex-end",
    // flex: 1
  }
});
