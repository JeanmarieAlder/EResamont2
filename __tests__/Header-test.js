import * as React from "react";
import renderer from "react-test-renderer";
import { render, fireEvent } from "react-native-testing-library";
import { LanguageContext } from "../shared/LanguageContext";
import Header from "../components/Header";
import utilities from "../utils/utilities";

let openDrawerCalled;
let navigation = {
  openDrawer: () => {
    console.log("open drawer");
    openDrawerCalled = true;
  },
  toggleDrawer: () => {
    console.log("toggle Drawer called");
  },
  navigate: () => {
    console.log("navigate called");
  }
};

let titleTab = [{ title: "title 1" }, { title: "title 2" }];

jest.mock("../utils/utilities");
utilities.findLanguageIndex = jest.fn(() => {
  return 1;
});
jest.mock("../routes/Drawer", () => "AppNavigator");

describe("Header", () => {
  const language = 1; //French
  const setLanguage = newLanguage => {
    this.language = newLanguage;
  };
  jest.useFakeTimers();

  it(`renders correctly`, () => {
    const tree = renderer
      .create(
        <LanguageContext.Provider value={{ language }}>
          <Header title={"test title"} />
        </LanguageContext.Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("opens menu when icon clicked", () => {
    openDrawerCalled = false;
    const { getByTestId } = render(
      <LanguageContext.Provider value={{ language, setLanguage }}>
        <Header title={"test title"} navigation={navigation} />
      </LanguageContext.Provider>
    );
    const element = getByTestId("header-button-menu");
    fireEvent.press(element);
    expect(openDrawerCalled).toBeTruthy();
  });
  it("finds title language if different from home or test title", () => {
    render(
      <LanguageContext.Provider value={{ language, setLanguage }}>
        <Header title={titleTab} navigation={navigation} />
      </LanguageContext.Provider>
    );
    expect(utilities.findLanguageIndex).toHaveBeenCalledTimes(1);
  });
});
