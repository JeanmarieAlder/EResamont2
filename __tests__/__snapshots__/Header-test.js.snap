// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders correctly 1`] = `
<View
  style={
    Object {
      "flexDirection": "row",
    }
  }
>
  <Text />
  <View>
    <Text
      style={
        Object {
          "color": "#232323",
          "fontFamily": "EuphemiaUCAS-Bold",
          "fontSize": 17,
          "fontWeight": "bold",
          "marginLeft": 5,
        }
      }
    >
      test title
    </Text>
  </View>
</View>
`;
